sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java

```diff
@@ -0,0 +1,289 @@
+        }
+
+        return throwable;
+    }
+
+    private String getMessageWithContext(String message) {
+        if (message == null) {
+            message = "";
+        }
+
+        int contextSize = context == null ? 0 : context.size();
+
+        StringBuilder sb = new StringBuilder(20 + contextSize * 20 + message.length());
+        sb.append("{\"")
+            // message must be first for log parsing tooling to work, key also works as a
+            // marker for Azure SDK logs so we'll write it even if there is no message
+            .append(AZURE_SDK_LOG_MESSAGE_KEY)
+            .append("\":\"")
+            .append(message)
+            .append("\",");
+
+        if (contextSize > 0) {
+            for (int i = 0; i < context.size() - 1; i++) {
+                context.get(i)
+                    .writeKeyAndValue(sb)
+                    .append(",");
+            }
+
+            // json does not allow trailing commas
+            context.get(context.size() - 1)
+                .writeKeyAndValue(sb);
+        }
+
+        sb.append("}");
+        return sb.toString();
+    }
+
+    private void addKeyValueInternal(String key, Object value) {
+        if (this.context == null) {
+            this.context = new ArrayList<>();
+        }
+
+        this.context.add(new ContextKeyValuePair(key, value));
+    }
+
+    /**
+     * Key value pair with basic serialization capabilities.
+     */
+    private static final class ContextKeyValuePair {
+        private final String key;
+        private final Object value;
+        private final Supplier<String> valueSupplier;
+
+        ContextKeyValuePair(String key, Object value) {
+            this.key = key;
+            this.value = value;
+            this.valueSupplier = null;
+        }
+
+        ContextKeyValuePair(String key, Supplier<String> valueSupplier) {
+            this.key = key;
+            this.value = null;
+            this.valueSupplier = valueSupplier;
+        }
+
+        /**
+         * Writes {"key":"value"} json string to provided StringBuilder.
+         */
+        public StringBuilder writeKeyAndValue(StringBuilder formatter) {
+            formatter.append("\"")
+                .append(key)
+                .append("\"")
+                .append(":");
+
+            String valueStr = null;
+            if (value != null) {
+                // LoggingEventBuilder only supports primitives and strings
+                if (!(value instanceof String)) {
+                    return formatter.append(value);
+                }
+
+                valueStr = (String) value;
+            } else if (valueSupplier != null) {
+                valueStr = valueSupplier.get();
+            }
+
+            if (valueStr == null) {
+                return formatter.append("null");
+            }
+
+            return formatter.append("\"")
+                .append(valueStr)
```

Comment:
great catch! I went to the rabbit hole there and had to change things a bit (e.g. we have to format message first before we escape it (because if formatting anchor `{}`)